test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,394 @@
+"""
+Tests for the ZooKeeper backup/restore scripts.
+"""
+
+import logging
+import subprocess
+import textwrap
+import uuid
+from pathlib import Path
+from typing import List
+
+import pytest
+from _pytest.fixtures import SubRequest
+from cluster_helpers import wait_for_dcos_oss
+from dcos_e2e.backends import Docker
+from dcos_e2e.cluster import Cluster
+from dcos_e2e.node import Output
+from kazoo.client import KazooClient
+from kazoo.exceptions import NoNodeError
+
+
+LOGGER = logging.getLogger(__name__)
+
+# Arbitrary value written to ZooKeeper.
+FLAG = b'flag'
+
+
+def _zk_set_flag(hosts: str, ephemeral: bool = False) -> str:
+    """
+    Store the `FLAG` value in ZooKeeper in a random Znode.
+    """
+    znode = '/{}'.format(uuid.uuid4())
+    zk_client = KazooClient(hosts=hosts)
+    zk_client.start()
+    zk_client.create(znode, makepath=True, ephemeral=ephemeral)
+    zk_client.set(znode, FLAG)
+    zk_client.stop()
+    return znode
+
+
+def _zk_check_flag(hosts: str, znode: str) -> None:
+    """
+    The `FLAG` value is found in ZooKeeper at the `znode` path.
+    """
+    zk_client = KazooClient(hosts=hosts, read_only=True)
+    zk_client.start()
+    value = zk_client.get(znode)
+    zk_client.stop()
+    assert value[0] == FLAG
+
+
+def _bash(cmd: List[str]) -> None:
+    """
+    Execute `cmd` in a subprocess, log potential errors.
+    """
+    LOGGER.debug('BASH ARGS: {}'.format(' '.join(cmd)))
+    try:
+        result = subprocess.check_output(
+            args=' '.join(cmd),
+            stderr=subprocess.STDOUT,
+            shell=True,
+        )
+        LOGGER.debug(result.decode())
+    except subprocess.CalledProcessError as exc:
+        LOGGER.error(exc.output.decode())
+        raise
+
+
+BACKUP_SCRIPT = textwrap.dedent(
+    """\
+#!/bin/bash -e
+
+DESTINATION_DIR=$PWD
+REMOTE_TMP_DIR=/tmp/zk_backup
```

Comment:
:+1: I guess you are aware that all shell expansion is done on the local host, so that pid will be used then.

Reply:
Yes. You could add the `$$` on the remote scripts, but that would make it difficult to `scp` the archive. So using the local pid is reasonable.

Reply:
There's some risk of a collision here. Either two backups running concurrently could share this tmp directory, or a failed command could leave a tmp directory for another run to find. 

Redefine `REMOTE_TMP_DIR` to be a shared temp directory (e.g. `/tmp` but configurable by user), and then create a `REMOTE_BACKUP_DIR=${REMOTE_TMP_DIR}/zkbackup.$$`. This incorporates the local pid and will have a better chance to be unique. The ssh script then creates the `REMOTE_BACKUP_DIR` and puts everything inside it.